Variables    
There are ten variables in the dataset as follows:

1. REF (reference number). The highest REF numbers were the last entries made. They are not unique values.
2. Company name or manufacturer   
3. Company location (Country)
4. Date of review of the chocolate ratings
5. Origin of bean (Country)
6. Specific bean origin or bar name
7. Cocoa percent
8. Ingredients - Represents the number of ingredients in the chocolate: (B = Beans, S = Sugar, S* = Sweetener other than white cane or beet sugar, C = Cocoa Butter, V = Vanilla, L = Lecithin, Sa = Salt)
9. Most memorable characteristics
10. Rating

Lecithin: a natural emulsifier found in egg yolks, soybeans, and sunflower seeds. It's added to chocolate to change its texture and consistency. It keeps the cocoa butter and cream from separating and improves its flow properties.

Rating Scale   
The ratings are between 1 and 5, with 1 considered the lowest rating and 5 as the highest rating possible.    

4.0 = Outstanding   
3.5 - 3.9 = Highly Recommended   
3.0 - 3.49 = Recommended   
2.0 - 2.9 = Disappointing   
1.0 - 1.9 = Unpleasant   


Import libraries:

In [3]:
import pandas as pd

df = pd.read_csv('chocolate_bar_ratings_2022.csv')
df


,REF,Company (Manufacturer),Company Location,Review Date,Country of Bean Origin,Specific Bean Origin or Bar Name,Cocoa Percent,Ingredients,Most Memorable Characteristics,Rating
0,2454,5150,U.S.A.,2019,Tanzania,"Kokoa Kamili, batch 1",76%,"3- B,S,C","rich cocoa, fatty, bready",3.25
1,2454,5150,U.S.A.,2019,Madagascar,"Bejofo Estate, batch 1",76%,"3- B,S,C","cocoa, blackberry, full body",3.75
2,2458,5150,U.S.A.,2019,Dominican Republic,"Zorzal, batch 1",76%,"3- B,S,C","cocoa, vegetal, savory",3.50
3,2542,5150,U.S.A.,2021,Fiji,"Matasawalevu, batch 1",68%,"3- B,S,C","chewy, off, rubbery",3.00
4,2542,5150,U.S.A.,2021,India,"Anamalai, batch 1",68%,"3- B,S,C","milk brownie, macadamia,chewy",3.50
...,...,...,...,...,...,...,...,...,...,...
2583,1205,Zotter,Austria,2014,Blend,Raw,80%,"4- B,S*,C,Sa","waxy, cloying, vegetal",2.75
2584,1996,Zotter,Austria,2017,Colombia,"APROCAFA, Acandi",75%,"3- B,S,C","strong nutty, marshmallow",3.75
2585,2036,Zotter,Austria,2018,Blend,"Dry Aged, 30 yr Anniversary bar",75%,"3- B,S,C","fatty, earthy, cocoa",3.00
2586,2170,Zotter,Austria,2018,Congo,Mountains of the Moon,70%,"3- B,S,C","fatty, mild nuts, mild fruit",3.25


Get an overview of the dataset in the next few cells:

In [7]:
df.head()


,REF,Company (Manufacturer),Company Location,Review Date,Country of Bean Origin,Specific Bean Origin or Bar Name,Cocoa Percent,Ingredients,Most Memorable Characteristics,Rating
0,2454,5150,U.S.A.,2019,Tanzania,"Kokoa Kamili, batch 1",76%,"3- B,S,C","rich cocoa, fatty, bready",3.25
1,2454,5150,U.S.A.,2019,Madagascar,"Bejofo Estate, batch 1",76%,"3- B,S,C","cocoa, blackberry, full body",3.75
2,2458,5150,U.S.A.,2019,Dominican Republic,"Zorzal, batch 1",76%,"3- B,S,C","cocoa, vegetal, savory",3.50
3,2542,5150,U.S.A.,2021,Fiji,"Matasawalevu, batch 1",68%,"3- B,S,C","chewy, off, rubbery",3.00
4,2542,5150,U.S.A.,2021,India,"Anamalai, batch 1",68%,"3- B,S,C","milk brownie, macadamia,chewy",3.50


In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2588 entries, 0 to 2587
Data columns (total 10 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   REF                               2588 non-null   int64  
 1   Company (Manufacturer)            2588 non-null   str    
 2   Company Location                  2588 non-null   str    
 3   Review Date                       2588 non-null   int64  
 4   Country of Bean Origin            2588 non-null   str    
 5   Specific Bean Origin or Bar Name  2588 non-null   str    
 6   Cocoa Percent                     2588 non-null   str    
 7   Ingredients                       2501 non-null   str    
 8   Most Memorable Characteristics    2588 non-null   str    
 9   Rating                            2588 non-null   float64
dtypes: float64(1), int64(2), str(7)
memory usage: 202.3 KB


In [9]:
df.shape

(2588, 10)

In [10]:
print(df.dtypes)

REF                                   int64
Company (Manufacturer)                  str
Company Location                        str
Review Date                           int64
Country of Bean Origin                  str
Specific Bean Origin or Bar Name        str
Cocoa Percent                           str
Ingredients                             str
Most Memorable Characteristics          str
Rating                              float64
dtype: object


These data types look appropriate. I might want to change the Cocoa Percent to an integer for plotting purposes.    

Next, I will check for null values.  There are 2588 rows, and the Ingredients column has 87 null values according to 
df.info() above. I need to find out which ones.

In [ ]:
df.isnull()

,Company (Manufacturer),Company Location,Review Date,Country of Bean Origin,Specific Bean Origin or Bar Name,Cocoa Percent,Ingredients,Most Memorable Characteristics,Rating
0,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...
2583,False,False,False,False,False,False,False,False,False
2584,False,False,False,False,False,False,False,False,False
2585,False,False,False,False,False,False,False,False,False
2586,False,False,False,False,False,False,False,False,False


Checking for null values. There are 2588 rows, and the Ingredients column has 87 null values according to 
df.info() above. Need to find out which ones.

In [11]:
df.isna()

,REF,Company (Manufacturer),Company Location,Review Date,Country of Bean Origin,Specific Bean Origin or Bar Name,Cocoa Percent,Ingredients,Most Memorable Characteristics,Rating
0,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...
2583,False,False,False,False,False,False,False,False,False,False
2584,False,False,False,False,False,False,False,False,False,False
2585,False,False,False,False,False,False,False,False,False,False
2586,False,False,False,False,False,False,False,False,False,False


This tells us that the missing values are in the ingredients column. It is the same thing as df.info()

In [12]:
df.isna().any()

REF                                 False
Company (Manufacturer)              False
Company Location                    False
Review Date                         False
Country of Bean Origin              False
Specific Bean Origin or Bar Name    False
Cocoa Percent                       False
Ingredients                          True
Most Memorable Characteristics      False
Rating                              False
dtype: bool

Find which rows are missing data in the Ingredients column.

In [13]:
null_rows = df[df['Ingredients'].isnull()]
print(null_rows)

       REF          Company (Manufacturer) Company Location  Review Date  \
88    1964                   Amazing Cacao           Russia         2017   
89    1964                   Amazing Cacao           Russia         2017   
167    486             Artisan du Chocolat             U.K.         2010   
303     81                          Bonnat           France         2006   
330    341          Bouga Cacao (Tulicorp)          Ecuador         2009   
...    ...                             ...              ...          ...   
2478   153  Vintage Plantations (Tulicorp)           U.S.A.         2007   
2479   153  Vintage Plantations (Tulicorp)           U.S.A.         2007   
2480   153  Vintage Plantations (Tulicorp)           U.S.A.         2007   
2505   733                      Whittakers      New Zealand         2011   
2515  1109                  Willie's Cacao             U.K.         2013   

     Country of Bean Origin   Specific Bean Origin or Bar Name Cocoa Percent  \
88     

Handling missing data  
1. Fill in the missing ingredients with the most common value.   
B, S, C (Beans, Sugar, Cocoa Butter) = 39.7%   
B, S (Beans, Sugar) = 28%  
B, S, C, L (Beans, Sugar, Cocoa Butter, Lecithin) =  11%  
These three unique value combinations together comprise 78.7% of the rows in the dataset.

2. Drop the rows with missing values. I would be dropping 3.3% of the dataset. 

3. Fill in the missing values with 'unknown' or 'not provided' for missing strings.

I will replace the missing values with the most common value (B, S, C) since I've already done visualizations on the full dataset. In the future, I would drop these rows. I decided to use the default (inplace=False) in order to return a copy where the replacing is done. I am concerned that using inplace=True to change the original DataFrame might interfere with the visualizations I've already created.

In [14]:
df_nonull = df.fillna('B, S, C')
df_nonull.head()

,REF,Company (Manufacturer),Company Location,Review Date,Country of Bean Origin,Specific Bean Origin or Bar Name,Cocoa Percent,Ingredients,Most Memorable Characteristics,Rating
0,2454,5150,U.S.A.,2019,Tanzania,"Kokoa Kamili, batch 1",76%,"3- B,S,C","rich cocoa, fatty, bready",3.25
1,2454,5150,U.S.A.,2019,Madagascar,"Bejofo Estate, batch 1",76%,"3- B,S,C","cocoa, blackberry, full body",3.75
2,2458,5150,U.S.A.,2019,Dominican Republic,"Zorzal, batch 1",76%,"3- B,S,C","cocoa, vegetal, savory",3.50
3,2542,5150,U.S.A.,2021,Fiji,"Matasawalevu, batch 1",68%,"3- B,S,C","chewy, off, rubbery",3.00
4,2542,5150,U.S.A.,2021,India,"Anamalai, batch 1",68%,"3- B,S,C","milk brownie, macadamia,chewy",3.50


Check again for missing values in the Ingredients column:

In [15]:
null_rows = df_nonull[df_nonull['Ingredients'].isnull()]
print(null_rows)

Empty DataFrame
Columns: [REF, Company (Manufacturer), Company Location, Review Date, Country of Bean Origin, Specific Bean Origin or Bar Name, Cocoa Percent, Ingredients, Most Memorable Characteristics, Rating]
Index: []


Assigning df_nonull back to df:

In [16]:
df = df_nonull
df.head

<bound method NDFrame.head of        REF Company (Manufacturer) Company Location  Review Date  \
0     2454                   5150           U.S.A.         2019   
1     2454                   5150           U.S.A.         2019   
2     2458                   5150           U.S.A.         2019   
3     2542                   5150           U.S.A.         2021   
4     2542                   5150           U.S.A.         2021   
...    ...                    ...              ...          ...   
2583  1205                 Zotter          Austria         2014   
2584  1996                 Zotter          Austria         2017   
2585  2036                 Zotter          Austria         2018   
2586  2170                 Zotter          Austria         2018   
2587  2170                 Zotter          Austria         2018   

     Country of Bean Origin Specific Bean Origin or Bar Name Cocoa Percent  \
0                  Tanzania            Kokoa Kamili, batch 1           76%   
1        

Checking again for null rows. It returned an empty dataframe for null values in the Ingredients column, meaning that there are no more NaN values.  

In [17]:
null_rows = df[df['Ingredients'].isnull()]
print(null_rows)

Empty DataFrame
Columns: [REF, Company (Manufacturer), Company Location, Review Date, Country of Bean Origin, Specific Bean Origin or Bar Name, Cocoa Percent, Ingredients, Most Memorable Characteristics, Rating]
Index: []


Checking again for any null values in the entire DataFrame:

In [18]:
df.isnull().sum()

REF                                 0
Company (Manufacturer)              0
Company Location                    0
Review Date                         0
Country of Bean Origin              0
Specific Bean Origin or Bar Name    0
Cocoa Percent                       0
Ingredients                         0
Most Memorable Characteristics      0
Rating                              0
dtype: int64

Check the number of occurrences of each unique combination of ingredients in the Ingredients column.

In [19]:
df['Ingredients'].value_counts()


Ingredients
3- B,S,C          1029
2- B,S             738
4- B,S,C,L         293
5- B,S,C,V,L       184
4- B,S,C,V         141
B, S, C             87
2- B,S*             31
4- B,S*,C,Sa        20
3- B,S*,C           13
3- B,S,L             8
4- B,S*,C,V          7
5-B,S,C,V,Sa         6
1- B                 6
4- B,S,V,L           5
4- B,S,C,Sa          5
6-B,S,C,V,L,Sa       4
3- B,S,V             3
4- B,S*,V,L          3
4- B,S*,C,L          2
3- B,S*,Sa           1
2- B,C               1
5- B,S,C,L,Sa        1
Name: count, dtype: int64

When trying to replace the NaN values in the Ingredients column, I replaced NaN with 'B, S, C' instead of '3- B,S,C' which created a new unique value. I wanted these added to the most common value, which is '3- B,S,C'. Now I will need to replace all instances of 'B, S, C' with '3- B,S,C'.  

In [20]:
df['Ingredients'] = df['Ingredients'].replace('B, S, C', '3- B,S,C')
print(df)

       REF Company (Manufacturer) Company Location  Review Date  \
0     2454                   5150           U.S.A.         2019   
1     2454                   5150           U.S.A.         2019   
2     2458                   5150           U.S.A.         2019   
3     2542                   5150           U.S.A.         2021   
4     2542                   5150           U.S.A.         2021   
...    ...                    ...              ...          ...   
2583  1205                 Zotter          Austria         2014   
2584  1996                 Zotter          Austria         2017   
2585  2036                 Zotter          Austria         2018   
2586  2170                 Zotter          Austria         2018   
2587  2170                 Zotter          Austria         2018   

     Country of Bean Origin Specific Bean Origin or Bar Name Cocoa Percent  \
0                  Tanzania            Kokoa Kamili, batch 1           76%   
1                Madagascar           B

Checking to make sure my changes worked

In [21]:
df['Ingredients'].value_counts()

Ingredients
3- B,S,C          1116
2- B,S             738
4- B,S,C,L         293
5- B,S,C,V,L       184
4- B,S,C,V         141
2- B,S*             31
4- B,S*,C,Sa        20
3- B,S*,C           13
3- B,S,L             8
4- B,S*,C,V          7
5-B,S,C,V,Sa         6
1- B                 6
4- B,S,V,L           5
4- B,S,C,Sa          5
6-B,S,C,V,L,Sa       4
3- B,S,V             3
4- B,S*,V,L          3
4- B,S*,C,L          2
3- B,S*,Sa           1
2- B,C               1
5- B,S,C,L,Sa        1
Name: count, dtype: int64

Cleaning up two other values by adding a space, for uniformity

In [22]:
df['Ingredients'] = df['Ingredients'].replace({'5-B,S,C,V,Sa': '5- B,S,C,V,Sa', '6-B,S,C,V,L,Sa': '6- B,S,C,V,L,Sa'})
print(df)

       REF Company (Manufacturer) Company Location  Review Date  \
0     2454                   5150           U.S.A.         2019   
1     2454                   5150           U.S.A.         2019   
2     2458                   5150           U.S.A.         2019   
3     2542                   5150           U.S.A.         2021   
4     2542                   5150           U.S.A.         2021   
...    ...                    ...              ...          ...   
2583  1205                 Zotter          Austria         2014   
2584  1996                 Zotter          Austria         2017   
2585  2036                 Zotter          Austria         2018   
2586  2170                 Zotter          Austria         2018   
2587  2170                 Zotter          Austria         2018   

     Country of Bean Origin Specific Bean Origin or Bar Name Cocoa Percent  \
0                  Tanzania            Kokoa Kamili, batch 1           76%   
1                Madagascar           B

Checking to see if changes were effective

In [23]:
df['Ingredients'].value_counts()

Ingredients
3- B,S,C           1116
2- B,S              738
4- B,S,C,L          293
5- B,S,C,V,L        184
4- B,S,C,V          141
2- B,S*              31
4- B,S*,C,Sa         20
3- B,S*,C            13
3- B,S,L              8
4- B,S*,C,V           7
5- B,S,C,V,Sa         6
1- B                  6
4- B,S,V,L            5
4- B,S,C,Sa           5
6- B,S,C,V,L,Sa       4
3- B,S,V              3
4- B,S*,V,L           3
4- B,S*,C,L           2
3- B,S*,Sa            1
2- B,C                1
5- B,S,C,L,Sa         1
Name: count, dtype: int64

Checking the number of entries for each country

In [24]:
df['Country of Bean Origin'].value_counts()


Country of Bean Origin
Venezuela             254
Peru                  248
Dominican Republic    234
Ecuador               223
Madagascar            184
                     ... 
China                   1
Principe                1
Sulawesi                1
Suriname                1
DR Congo                1
Name: count, Length: 63, dtype: int64

Checking value counts for review date

In [25]:
df['Review Date'].value_counts()

Review Date
2015    284
2014    247
2018    228
2016    217
2012    194
2019    193
2013    183
2021    175
2011    163
2009    123
2010    110
2017    105
2008     92
2020     81
2007     73
2006     62
2022     58
Name: count, dtype: int64

Top ten bean-producing countries

In [26]:
df['Country of Bean Origin'].value_counts().head(10)

Country of Bean Origin
Venezuela             254
Peru                  248
Dominican Republic    234
Ecuador               223
Madagascar            184
Blend                 156
Nicaragua             103
Tanzania               81
Bolivia                81
Brazil                 81
Name: count, dtype: int64

Checking the number of occurrences of each manufacturer

In [27]:
df['Company (Manufacturer)'].value_counts()

Company (Manufacturer)
Soma                    57
Fresco                  39
Arete                   32
Bonnat                  30
A. Morin                29
                        ..
Vietcacao (A. Morin)     1
Vintage Plantations      1
Whittakers               1
Xocolat                  1
Zacharias                1
Name: count, Length: 593, dtype: int64

Top ten chocolate manufacturers by number of bars produced

In [28]:
df['Company (Manufacturer)'].value_counts().head(10)

Company (Manufacturer)
Soma         57
Fresco       39
Arete        32
Bonnat       30
A. Morin     29
Dandelion    25
Pralus       25
Domori       24
Guittard     22
Valrhona     22
Name: count, dtype: int64

Checking the number of countries the various chocolate-manufacturing companies are in

In [29]:
df['Company Location'].value_counts()

Company Location
U.S.A.                   1168
France                    179
Canada                    178
U.K.                      134
Italy                      79
                         ... 
Wales                       1
Martinique                  1
St.Vincent-Grenadines       1
Ghana                       1
Suriname                    1
Name: count, Length: 65, dtype: int64

Top 10 chocolate-producing countries

In [30]:
df['Company Location'].value_counts().head(10)

Company Location
U.S.A.         1168
France          179
Canada          178
U.K.            134
Italy            79
Belgium          72
Ecuador          58
Australia        53
Switzerland      44
Germany          43
Name: count, dtype: int64

Number of occurrences for each unique combination of characteristics

In [31]:
df['Most Memorable Characteristics'].value_counts()

Most Memorable Characteristics
spicy, cocoa                    4
creamy, nutty, cocoa            4
sweet, vanilla, cocoa           3
sweet, marshmallow              3
sandy, sweet, cocoa             3
                               ..
waxy, cloying, vegetal          1
strong nutty, marshmallow       1
fatty, earthy, cocoa            1
fatty, mild nuts, mild fruit    1
muted, roasty, accessible       1
Name: count, Length: 2545, dtype: int64

In [32]:
df['Most Memorable Characteristics'].nunique

<bound method IndexOpsMixin.nunique of 0           rich cocoa, fatty, bready
1        cocoa, blackberry, full body
2              cocoa, vegetal, savory
3                 chewy, off, rubbery
4       milk brownie, macadamia,chewy
                    ...              
2583           waxy, cloying, vegetal
2584        strong nutty, marshmallow
2585             fatty, earthy, cocoa
2586     fatty, mild nuts, mild fruit
2587        muted, roasty, accessible
Name: Most Memorable Characteristics, Length: 2588, dtype: str>

There are 2588 unique values in the 'Characteristics' column (I did this after I shortened the column names)  and 2588 rows in the entire dataset. This means that each entry has a unique combination of characteristics.  I wanted to know this for my ERD in Lucidchart.   

Top ten combinations of flavor characteristics

In [33]:
df['Most Memorable Characteristics'].value_counts().head(10)

Most Memorable Characteristics
spicy, cocoa             4
creamy, nutty, cocoa     4
sweet, vanilla, cocoa    3
sweet, marshmallow       3
sandy, sweet, cocoa      3
sandy, sweet             3
cherry, rich choco       3
sandy, roasty, nutty     2
creamy, fruit, cocoa     2
nutty, green             2
Name: count, dtype: int64

How to find out the 10-20 most frequently-occurring words?

In [35]:
df['Rating'].value_counts()

Rating
3.50    584
3.00    534
3.25    474
2.75    339
3.75    306
2.50    169
4.00    115
2.00     33
2.25     17
1.50     10
1.00      4
1.75      3
Name: count, dtype: int64

In [36]:
df['Cocoa Percent'].value_counts()

Cocoa Percent
70%       1070
75%        312
72%        309
65%         93
80%         90
68%         75
74%         69
73%         67
60%         46
71%         44
77%         43
85%         40
64%         37
76%         35
67%         35
66%         28
78%         22
100%        21
82%         18
55%         16
62%         16
63%         15
69%         13
90%          9
88%          8
58%          8
61%          7
81%          6
83%          5
72.50%       4
84%          4
91%          3
73.50%       2
79%          2
71.50%       2
56%          2
89%          2
99%          2
50%          1
86%          1
42%          1
60.50%       1
57%          1
46%          1
53%          1
87%          1
Name: count, dtype: int64

In [37]:
df.describe()

,REF,Review Date,Rating
count,2588.000000,2588.000000,2588.000000
mean,1459.185471,2014.545209,3.198319
std,773.852201,4.082722,0.444069
min,5.000000,2006.000000,1.000000
25%,820.000000,2012.000000,3.000000
50%,1470.000000,2015.000000,3.250000
75%,2122.000000,2018.000000,3.500000
max,2768.000000,2022.000000,4.000000


Change Cocoa Percent from string to integer
Remove the % sign

In [39]:
df['Cocoa Percent'] = df['Cocoa Percent'].str.replace('%', '')


This worked previously. I'm not sure why it's giving an error message now. OK, I see why now (I think). It has already changed it to a float, so I think that Python thinks this bit of code is no longer needed.  

Looks like it successfully changed it to a float in the cell below:

In [40]:
df['Cocoa Percent'] = df['Cocoa Percent'].astype('float')
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2588 entries, 0 to 2587
Data columns (total 10 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   REF                               2588 non-null   int64  
 1   Company (Manufacturer)            2588 non-null   str    
 2   Company Location                  2588 non-null   str    
 3   Review Date                       2588 non-null   int64  
 4   Country of Bean Origin            2588 non-null   str    
 5   Specific Bean Origin or Bar Name  2588 non-null   str    
 6   Cocoa Percent                     2588 non-null   float64
 7   Ingredients                       2588 non-null   str    
 8   Most Memorable Characteristics    2588 non-null   str    
 9   Rating                            2588 non-null   float64
dtypes: float64(2), int64(2), str(6)
memory usage: 202.3 KB


In [41]:
df['Cocoa Percent'].head()

0    76.0
1    76.0
2    76.0
3    68.0
4    68.0
Name: Cocoa Percent, dtype: float64

Check to see if it's treating this column as actual numerical values:

In [42]:
df['Cocoa Percent'].sort_values(ascending=True)

587      42.0
1451     46.0
112      50.0
2077     53.0
246      55.0
        ...  
2515    100.0
2135    100.0
2136    100.0
167     100.0
1939    100.0
Name: Cocoa Percent, Length: 2588, dtype: float64

In [43]:
df['Rating'].sort_values(ascending=True)

1719    1.0
436     1.0
601     1.0
641     1.0
779     1.5
       ... 
32      4.0
24      4.0
19      4.0
16      4.0
2152    4.0
Name: Rating, Length: 2588, dtype: float64

Checking another way to see if it's treating this column as actual numerical values:

In [44]:
df['Cocoa Percent'].describe()

count    2588.000000
mean       71.613022
std         5.577118
min        42.000000
25%        70.000000
50%        70.000000
75%        74.000000
max       100.000000
Name: Cocoa Percent, dtype: float64

In [45]:
print(df['Specific Bean Origin or Bar Name'])

0                 Kokoa Kamili, batch 1
1                Bejofo Estate, batch 1
2                       Zorzal, batch 1
3                 Matasawalevu, batch 1
4                     Anamalai, batch 1
                     ...               
2583                                Raw
2584                   APROCAFA, Acandi
2585    Dry Aged, 30 yr Anniversary bar
2586              Mountains of the Moon
2587                           Maya Mtn
Name: Specific Bean Origin or Bar Name, Length: 2588, dtype: str


I dropped the REF column because I didn't think it had any relevance to the analysis. It appears to be just a number that specifies when the entry was made. The highest REF numbers were the last entries made.     

Correction: I decided not to drop it because I am using it to connect tables in my Entity Relationship Diagram.  I removed df.drop code that I had in a code cell just beneath this one.

Check for duplicates:

In [46]:
df.duplicated()


0       False
1       False
2       False
3       False
4       False
        ...  
2583    False
2584    False
2585    False
2586    False
2587    False
Length: 2588, dtype: bool

Show duplicate rows, since above output is truncated. 
In this case, there are no duplicate rows. It returned an empty DataFrame.

In [ ]:
duplicate_rows = df[df.duplicated()]
print(duplicate_rows)

Empty DataFrame
Columns: [Company (Manufacturer), Company Location, Review Date, Country of Bean Origin, Specific Bean Origin or Bar Name, Cocoa Percent, Ingredients, Most Memorable Characteristics, Rating]
Index: []


Using .loc and .iloc:

In [ ]:
df.loc[9: 37]

,Company (Manufacturer),Company Location,Review Date,Country of Bean Origin,Specific Bean Origin or Bar Name,Cocoa Percent,Ingredients,Most Memorable Characteristics,Rating
9,A. Morin,France,2013,Panama,Panama,70.0,"4- B,S,C,L","brief fruit note, earthy, nutty",2.75
10,A. Morin,France,2013,Madagascar,"Madagascar, Criollo",70.0,"4- B,S,C,L","sticky, red fruit, sour",3.00
11,A. Morin,France,2013,Brazil,Brazil,70.0,"4- B,S,C,L",mild tobacco,3.25
12,A. Morin,France,2013,Ecuador,Equateur,70.0,"4- B,S,C,L","sandy, nutty, cocoa, fig",3.75
13,A. Morin,France,2013,Colombia,Colombie,70.0,"4- B,S,C,L","burnt rubber,alkalyzed notes",2.75
14,A. Morin,France,2013,Burma,Birmanie,70.0,"4- B,S,C,L","sticky, smokey, grass",3.00
15,A. Morin,France,2013,Papua New Guinea,Papua New Guinea,70.0,"4- B,S,C,L","mild fruit, strong smoke",3.25
16,A. Morin,France,2013,Venezuela,Chuao,70.0,"4- B,S,C,L","oily, nut, caramel, raspberry",4.00
17,A. Morin,France,2013,Peru,Piura,70.0,"4- B,S,C,L","green, nutty, cocoa",3.25
18,A. Morin,France,2013,Peru,Chanchamayo Province,70.0,"4- B,S,C,L","cocoa,sour,intense tangerine",3.50


Using .loc and .iloc

In [ ]:
df.iloc[10, 4]

'Madagascar, Criollo'

I want to shorten some of the column names for improved readability.

First, I run print(df.columns) to review my column names, so I don't have to scroll up and find them.  Then I rename the columns with df.rename(columns) with a dictionary of the old column names as the keys and the new column names as the values.

In [47]:
print(df.columns)

Index(['REF', 'Company (Manufacturer)', 'Company Location', 'Review Date',
       'Country of Bean Origin', 'Specific Bean Origin or Bar Name',
       'Cocoa Percent', 'Ingredients', 'Most Memorable Characteristics',
       'Rating'],
      dtype='str')


In [48]:
df.rename(columns={'Company (Manufacturer)': 'Company', 'Country of Bean Origin': 'Bean Origin', 'Specific Bean Origin or Bar Name': 'Bar Name', 'Cocoa Percent': 'Cocoa %', 'Most Memorable Characteristics': 'Characteristics'}, inplace=True)

df.head()

,REF,Company,Company Location,Review Date,Bean Origin,Bar Name,Cocoa %,Ingredients,Characteristics,Rating
0,2454,5150,U.S.A.,2019,Tanzania,"Kokoa Kamili, batch 1",76.0,"3- B,S,C","rich cocoa, fatty, bready",3.25
1,2454,5150,U.S.A.,2019,Madagascar,"Bejofo Estate, batch 1",76.0,"3- B,S,C","cocoa, blackberry, full body",3.75
2,2458,5150,U.S.A.,2019,Dominican Republic,"Zorzal, batch 1",76.0,"3- B,S,C","cocoa, vegetal, savory",3.50
3,2542,5150,U.S.A.,2021,Fiji,"Matasawalevu, batch 1",68.0,"3- B,S,C","chewy, off, rubbery",3.00
4,2542,5150,U.S.A.,2021,India,"Anamalai, batch 1",68.0,"3- B,S,C","milk brownie, macadamia,chewy",3.50


I need to save the dataframe as a "cleaned" csv file. Otherwise, I would be using the original csv file in my visualizations, which is what I think I did. Once I create the new cleaned csv file, I will have to load it into the notebook that contains my visualizations in place of the original csv file and rerun all the cells.  

In [6]:
df.to_csv('cleaned_chocolate_bar_ratings_2022.csv', index=False)

It did not save the cleaned version of my dataframe. Also, I tried to save it in the data folder but it gave me an error message.  So I just moved the file into the data folder in file explorer.        

I started at the top of the notebook and ran each cell again, thinking that it just needed a reset.  I will try to save again.

In [49]:
df.to_csv('chocolate_bar_ratings_2022_cleaned_1.csv', index = False)

It worked this time. I successfully saved the cleaned version of the dataframe.  After I moved it into my data folder, I changed the name to chocolate_bar_ratings_2022_cleaned.csv.  